### PR TITLE
enhancement(vrl): read grok aliases from file

### PR DESCRIPTION
**Pull request by itkovian**
This PR aims to address (in part) vectordotdev/vrl#89, allowing GROK aliases to be defined in one or more files. I am not sure 

- Files are assumed to contain JSON, with a single outer Object.
- Duplicate patterns are not handled
- Needs proper tests.

### CODE REVIEW

File: lib/vrl/compiler/src/function.rs [hunk 1]
```diff
@@ -535,13 +539,14 @@ impl diagnostic::DiagnosticMessage for Error {
             InvalidArgument { .. } => 403,
             ExpectedFunctionClosure => 420,
             ReadOnlyMutation { .. } => 315,
+            InvalidAliasSource { .. } => 401,
```
Comment (itkovian) [inline lib/vrl/compiler/src/function.rs:542]: Not sure this is the right value.

File: lib/vrl/stdlib/src/parse_groks.rs [hunk 2]
```diff
@@ -141,6 +146,33 @@ impl Function for ParseGroks {
...
+                    .try_bytes_utf8_lossy()
+                    .expect("filename not bytes")
+                    .into_owned();
+                Ok(path)
+            })
+            .collect::<std::result::Result<Vec<String>, vrl::function::Error>>()?;
+
+        for src in alias_sources {
+            let path = Path::new(&src);
+            let file = File::open(&path).unwrap();
```
Comment (itkovian) [inline lib/vrl/stdlib/src/parse_groks.rs:169]: This needs to be handled properly and an InvalidAliasSource should be returned, but I'm not sure what the proper way to do this is.

**Comment by jszwedko** [comment]
Thanks @itkovian ! The approach here seems sane. Just let us know when you want a more thorough review or if you have any questions.

**Comment by JeanMertz** [comment]
Hey @itkovian, thank you for this contribution!

We discussed this internally, and have a few concerns and thoughts around file-system access in VRL.

To start off, we currently don't support any user-configurable file-system access (outside of enrichment tables, but access to those are configured inside Vector, not in VRL). The reason we haven't done so yet, is because we want to make sure we get this right.

My initial thought on how we would like to provide file-system access is through a new static string type specifically for paths to files/directories. For example, one could write `fs'/foo/bar.json'` as a VRL static value, which would allow us to:

1. specifically define file-system path requirements for function arguments, but more importantly:
2. make file system access a compiler-implemented solution, which would allow the compiler to check certain invariants at boot-time (file exists, is accessible, and potentially in the future "is within the accessible range of sandboxed paths").
3. It would make function implementations easier, as they can assume that any potential errors have been handed by the compiler.

This is just one idea on how to do this, there might be others.

Given these higher-level questions, I think the best way forward is to first write an RFC to discuss this issue and figure out how we can resolve this problem.

Would you be up for writing such an RFC (we can guide you through it)? If not, we can leave this PR open for now, to gather more input/thoughts before someone finds the time to write an actual RFC to move this work forward.

Meanwhile, if you have an urgent need for a function like this, you can always compile Vector yourself, with this function included. I understand that's a less desired solution, but it would at least unblock you, if need be.

**Comment by itkovian** [comment]
Hi @JeanMertz, I am certainly willing to put in effort into writing something, but I am not (yet) familiar with the vector codebase, so help would certainly be appreciated. 

I already compiled this myself and have it running for a proof-of-concept at work (there are some other issues there I need to deal with, but nothing related to this). How should I start/proceed/communicate about this?

**Comment by JeanMertz** [comment]
@itkovian, thank you for your patience.

We can do this in incremental steps. The first step would be to write an RFC with the proposal laid out for others to comment on, and refine until we're all happy with the proposed solution.

The first step could be to read up on [existing RFCs](https://github.com/vectordotdev/vector/tree/36fb42b331b0a8ba0f230a272b9ffdb481f997dc/proto) we've published in the past. You can look at both VRL and non-VRL RFCs to get a sense of the goal of RFCs and how we structure them. You can also [search for PRs](https://github.com/vectordotdev/vector/pulls?q=is%3Apr+sort%3Aupdated-desc+label%3A%22domain%3A+rfc%22+) in which we propose RFCs to see how the comment process usually goes.

Once you've familiarized yourself with the process and expectations, you can create a copy of the [RFC template](https://github.com/vectordotdev/vector/blob/36fb42b331b0a8ba0f230a272b9ffdb481f997dc/rfcs/_YYYY-MM-DD-issue%23-title.md), and write an initial draft for this proposal. Please don't hesitate to leave blanks in the initial draft, and ask for help filling in those blanks (for example, in a draft PR) after posting your initial in-progress RFC.

**Comment by fuchsnj** [comment]
I apologize for the inconvenience, but VRL has been moved to a [separate repo](https://github.com/vectordotdev/vrl), and this PR will need to be moved over there.

**Comment by itkovian** [comment]
That's ok :)

**Pull request closed without merging**